### ISSUE TITLE
Bids and asks cannot be swapped

### ISSUE THREAD

**Issue by dimitri-xyz** [gnosis/dex-research#9]
Excellent work!

There are a few problems with the definitions of buy and sell orders that I think are causing confusion. I will open a few issues. The definitions are asymmetric and have lead to a mistake here:

https://github.com/gnosis/dex-research/blob/13e8ebac59fbc14a29de3ec5dd893b68991bb807/BatchAuctionOptimization/batchauctions.tex#L260-L264

Indeed a buyer "should not object" to receiving more tokens than he asked for, but only if those tokens come at *zero* cost. This is because the utility of these extra tokens can be zero, whereas the utility of the tokens used for the purchase may not. Thus, we may not conclude that we can spend any more $ \tau_k $ than strictly necessary. And it is *not* possible to do the translation.

Example: If I place an order to buy 2 Bitcoin for $6000 each. I indeed *must* have $12000 available. However, it is *not* the case, that I would prefer to buy 20,000 Bitcoins for the same $12000 rather than 19,999 bitcoins for $11999. I saved a dollar on the second alternative and the extra bitcoin is of no use to me.

This implies that it is not possible to swap the \overline{x} and \overline{y} maximum limits as described because the two orders represent distinct preferences (they mean different things).

I encountered the same problem when trying to design market orders for batch auctions.

**Comment by bh2smith** [comment]
Hey @tomtiger87, Have you had a chance to look into this?

**Comment by tomtiger87** [comment]
Hey @dimitri-xyz ,

thanks for reading and commenting on the document! I've seen your issues only yesterday (thanks to @bh2smith ).

I totally agree with you on the slight semantic difference between buy and sell orders. In the current version of the document which I have just merged in commit 8f3282c8, I am using only sell orders in all models in order to simplify notation and make the document easier to read. I have rewritten the paragraph that you had cited as follows:
https://github.com/gnosis/dex-research/blob/8f3282c81b777849fb5ed7d2f296ab5ab0ae28f4/BatchAuctionOptimization/batchauctions.tex#L238-L248

**Comment by dimitri-xyz** [comment]
Hi @tomtiger87,

I don't understand what you mean by "assuming that the trader has constant utility for an arbitrary number of buy-tokens". I assume this means that the amount of extra utility the trader gets from each extra token is constant. If this is the case, I can see the argument follow.

If you meant that the trader gets zero utility if he gets any other quantity, except for the arbitrary amount he wants, then I am not sure the argument follows unless this is a "fill-or-kill" order. But I am also not sure he should place the buy order in the first place. I'd have to think about it some more.

The problem is that usual market mechanism we use is not symmetric.

For example: In the BTC-USD market, a buy order could be limited by the USD funds used, the volume of BTC obtained or the price paid (it's a triple (cost,price,vol) but the cost usually cannot be specified). We can usually deduce an (upper/lower) bound from the other 2 values, but I don't see anything wrong with placing an order to buy 3 BTC for a price of at most $1000 USD/BTC but paying a total of at most 2500 USD. (Note that this order can be totally filled and the upper bound here is useless.) Once we see orders being specified in triples, it becomes easier to flip bids to asks and vice-versa.

**Comment by tomtiger87** [comment]
Hi @dimitri-xyz ,

thanks again! 

With "constant utility for an arbitrary number of buy-tokens" I indeed mean that every extra unit of the buy token has the same utility to the trader. Will try to rephrase in a next version.

Thanks also for the example on the cost-bounded order type, makes sense to me now! I just took a brief look into modelling such an order type within our combinatorial optimization approach and it looks like this order type would just add one more linear constraint (per order). Will keep that in mind for future versions..

**Comment by dimitri-xyz** [comment]
You're welcome @tomtiger87 !

Let's think of the USD/BTC market w.l.o.g. to make things simpler.

I think these cost bounds are specially confusing when converting bids and asks by swapping the tokens. When I place a bid to buy 3 BTC at $3000 USD/BTC each, I have established a volume target in BTC (of how much I want to get) and a price target in USD/BTC. However, when I place an ask to sell 3 BTC at $3000 USD/BTC, I have established a cost target in BTC (of how much I want to pay) and a price target (but not a volume target). The straight forward conversion of this ask into a bid is to buy as many dollars as possible for at most $ 1/3000 BTC/USD (a price target) and spending at most 3 BTC (a cost target).

Notice that of the 3 possible bounds (cost, price, volume), standard bids specify (explicitly) volume in BTC and price in USD/BTC whereas a standard asks *when seen as a bid* would specify cost in BTC and price in BTC/USD. In other words, when we convert the ask into a bid, we get a different kind of bid (one that specifies a target cost rather than a target volume). We can then use simplifying assumptions to make the two bids look the same, but in some sense we are already forced to deal with the 3 bounds.

(N.B. If you intent to later model market orders, market bids are usually modeled by limits with infinite price, so you likely will have to deal with all 3 bounds then.)
